### ISSUE TITLE
FindReferencesAsync returns false positive when class has explicit interface method implementation

### ISSUE THREAD

**Issue by Stevie-O** [dotnet/roslyn#34568]
**Version Used**: 
Microsoft.Build.Framework, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Microsoft.CodeAnalysis, Version=2.10.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
Microsoft.CodeAnalysis.Workspaces, Version=2.10.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
Microsoft.CodeAnalysis.CSharp.Workspaces, Version=2.10.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
Microsoft.CodeAnalysis.CSharp, Version=2.10.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
Microsoft.CSharp, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
**Steps to Reproduce**:

Take this code:
```C#
public interface IDriveable { void Drive(); }
public class Vehicle : IDriveable { public virtual void Drive() {} }
public class Motorcycle : Vehicle, IDriveable { public override void Drive() {} void IDriveable.Drive() { } }
public class Tests
{
	public static void TestMotorcycle()
	{
		var m = new Motorcycle();
		m.Drive();
	}
	public static void TestDriveable(IDriveable d)
	{
		d.Drive();
	}
}
```
and 

(A) Call `FindReferencesAsync` for `Motorycle.Drive()`.
(B) Call `FindReferencesAsync` for `IDriveable.Drive()`.

**Expected Behavior**:

(A) `IDriveable.Drive()` (`Tests.TestDriveable`) is not in the returned definitions
(B) `Motorcycle.Drive()` (`Tests.TestMotorcycle`) is not in the returned definitions

**Actual Behavior**:

They are.

For (A), the explicit interface method implementation means that no callsite that calls IDriveable.Drive can invoke Motorycle.Drive at runtime.

For (B), I'm not even sure what the semantics are supposed to be here, but again: in no way is a call to Motorcycle.Drive() relates to IDriveable.Drive.  The only argument one might possibly make to the contrary is that if you have a new class that inherits from Motorcycle and overrides Drive(), then a callsite for Motorcycle.Drive *can* invoke a method that also implements IDriveable.Drive.

**Comment by Stevie-O** [comment]
I couldn't attach it directly, so here's a gist of a LINQPad script showing the issue:
https://gist.github.com/Stevie-O/06ac50f4f219d2e1e50e5eabd200ed4b

**Comment by CyrusNajmabadi** [comment]
These are all considered cascaded definitions.  The order of cascading happens because IDrivable.Drive is linked to Vehicle.Drive in `class Vehicle`.  `Motorcyle.Drive` is then cascaded from this due to it overriding `Vehicle.Drive`.

Note that Motorcycle.Drive is trivially invokable through Vehicle.Drive.

--

it's not clear why any of this is really a problem either.  If your code is structured this way, it seems fine to indicate there are cascaded relations here.

**Comment by Stevie-O** [comment]
> These are all considered cascaded definitions. The order of cascading happens because IDrivable.Drive is linked to Vehicle.Drive in class Vehicle. Motorcyle.Drive is then cascaded from this due to it overriding Vehicle.Drive.

Yeah, you just described how FindReferencesAsync behaves.  My point is that _that behavior is not correct_.

> Note that Motorcycle.Drive is trivially invokable through Vehicle.Drive.

While this is true, it's also irrelevant. There's only one method in my example that comes close to invoking Vehicle.Drive, and that's Tests.TestMotorcycle; Tests.TestDriveable does _not_ invoke Vehicle.Drive.

It's a problem because I'm trying to do dependency/call-flow analysis, and FindReferences is giving me incorrect answers.  I'm trying to answer the question: "What are all call sites that can possibly call this method?".  

The documentation for FindReferences simply says "Finds all references to a symbol throughout a solution".  While that is strictly true, it's misleading, because it is also finding things that are *not* references to the specified symbol.  And now I'm stuck with the extremely thorny issue of figuring out which is which.

For non-virtual method invocations (note: this includes `base.Foo()`, which invokes a virtual method non-virtually), the problem trivial.

For virtual invocations, the problem is significantly more complex -- and FindReferences pretends it's not.
Granted, there's no guaranteed way to get it down to zero false positives in all cases (doing so in the general case is tantamount to solving the halting problem), but level of transitivity assumed is _way_ too high.

Here are things that FindReferences tells me are possible that are not:
- Interface method calls are being reported as related to unrelated methods that are not implement the interface (the problem I stated, because it was the most obviously incorrect)
- This case (which is giving me the most grief at the moment):
```
public class Vehicle { public virtual void Drive() { } }
public class Car : Vehicle { public override void Drive() {} }
public class Truck : Vehicle { public override void Drive() {} }
public class Tests
{
	public static void TestTruck()
	{
		var t = new Truck();
		// FindCallersAsync will list this as a possible call site for Car.Drive, which is ludicrous
		t.Drive();
	}
}
```
As the comment says, FindCallersAsync will claim that Car.Drive can be invoked through a Truck!

- This case which is also giving me trouble:
```
class A { public virtual void Foo() { } }
class B : A { public override void Foo() { } }
class C1 : B { public override void Foo() { } }
class C2 : B { public override void Foo() { } }
class BHelper
{
	public static void Foo(B b)
	{
		b.Foo();
	}
}
```
It is obvious that BHelper.Foo cannot call A.Foo -- but FindReferences claims it is possible!

Again, I don't expect FindReferences/FindCallersAsync to be perfect.  I wasn't even going to file an issue for the latter two problems until I had a clear sense of a fix for them.  But the interface thing seemed obvious (and I figured that the fix for that would give me a clue as how to fix the other two.)

**Comment by CyrusNajmabadi** [comment]
> It's a problem because I'm trying to do dependency/call-flow analysis

FindReferenceAsync doesn't attempt to solve that.  If you want that, you'll have to do additional analysis.  What it does is find references to the actual symbl you asked for.  It also cascades to useful additional other symbols in case the values are desired.  This is done as a perf optimization as manually cascading and re-finding would be very expensive to do externally.

Different clients of the FindReferences api have to make determinations about which cascaded values to use and if they're appropriate.  

>  but level of transitivity assumed is way too high.

The level is too high for your needs.

NOte: the example you gave seems somewhat strange as well.  I get that you're concerned about thsi case.  But what are the chances of this case actually happening in practice?  Having someone override an interface impl and also provide an explicit impl of that same interface member seems incredibly esoteric.  Given the chance of this happening in practice, it seems not really that interesting to care so much about.  And, if it is super important for you, you could always check for this and filter out these results.

**Comment by CyrusNajmabadi** [comment]
> ```c#
> 	// FindCallersAsync will list this as a possible call site for Car.Drive, which is ludicrous
>		t.Drive();
> ```

That should not happen.  That should only be listed as a site for Truck.Drive.  You can even see that in the UI:

![image](https://user-images.githubusercontent.com/4564579/55203630-57fbcd00-5189-11e9-9d15-e6e7d2d5560f.png)

As you can see, it's not a site for Car.Drive.

**Comment by CyrusNajmabadi** [comment]
> It is obvious that BHelper.Foo cannot call A.Foo -- but FindReferences claims it is possible!

You are not understanding the results.  It's not saying that A.Foo is called here (and, if you look at the results, you'll see no hits to A.Foo inside BHelper.Foo.  What it's saying is "`B.Foo` is related to `A.Foo` (since there is a virtua/override connection), since you want the information about references to `B.Foo`, we're also going to give you the information about `A.Foo`".  That's it.

Note that for your needs this isn't helpful.  BUt other services need it.  For example, 'Rename' would certainly want to rename all these `Foo` methods if one of them changed.  That's why the information is included, so it can be a superset of valuable information provided to any of the many myriad features and services that sit on top of this.

**Comment by CyrusNajmabadi** [comment]
AFAICT, FindReferences is behaving correctly.  It's totally possible that FindCallers is not.  Looks like that's some sort of helper that just tries to combine all the FindRef data (i.e. direct refs and cascaded refs) into one big bag.  My recommendation woudl be to just not use that API as i doubt it's been vetted for this sort of thing (and changing it would likely break someone).

**Comment by Stevie-O** [comment]
Okay, so the issue is that FindCallersAsync (which is the method I'm actually using) is poorly designed.  I've been focusing on FindReferencesAsync because FindCallers uses FindReferences to get its data, then rewrites things.

This all leads back to: The intended behavior is _very_ poorly documented, and it's difficult to infer the correct behavior, because there are very few examples for using Roslyn with SymbolFinder (most examples are focused on Visual Studio code analysis plugins, which generally focus on a single Compilation unit.)

**Comment by CyrusNajmabadi** [comment]
> This all leads back to: The intended behavior is very poorly documented, and it's difficult to infer the correct behavior, because there are very few examples for using Roslyn with SymbolFinder (most examples are focused on Visual Studio code analysis plugins, which generally focus on a single Compilation unit.)

I'm sure PRs with updated docs would be welcome :)

**Comment by Stevie-O** [comment]
How does this sound?

FindReferencesAsync considers the following symbols to be related:
- A symbol is related to itself.
- Symbols (methods, properties, events) providing implementations with the same virtual symbol are related.
  - Symbols that implement an interface member are related to that interface member.
  - Symbols that override a base class member are related to that base class member.

FindReferencesAsync computes the transitive closure of symbols related to <param ref="symbol" /> and, for each such symbol, returns a <see cref="ReferencedSymbol" /> instance for that symbol that contains a the spans of the syntax nodes that refer to this symbol.

(NOTE: In all of my experiments, this turns out to be an IdentifierNameSyntax node.  Is it even possible to get a different node type here, at least in C#?)

**Comment by CyrusNajmabadi** [comment]
>FindReferencesAsync considers the following symbols to be related:

yes, that's pretty much spot on.  there may be some tiny additional details there, but it would involve deeply examining the impl.  For example, both parts of a 'partial' method will be included.

> (NOTE: In all of my experiments, this turns out to be an IdentifierNameSyntax node. Is it even possible to get a different node type here, at least in C#?)

Yup, definitely possible.  For example, if you are searching for things like `IEnumerable<X>.GetEnumerator()` we'll find hits inside `foreach` statements.  We'll either return the span of the foreach-node, or at least hte location of the `foreach` token.

**Comment by jinujoseph** [comment]
Related to #32738

**Comment by CyrusNajmabadi** [comment]
Closing out as this is working as intended.
